app/routes.py

```diff
@@ -1,2 +1,273 @@
+
+        return make_response({
+            "task":{ 
+                "id": new_task.task_id,
+                "title": new_task.title,
+                "description": new_task.description,
+                "is_complete": completed_at
+            }
+        }), 201   
+
+# return one task
+@task_bp.route("/<task_id>", methods=["GET", "PUT", "DELETE"], strict_slashes = False)
+def handle_task(task_id):
+    task = Task.query.get(task_id)
+
+    if request.method == "GET":
+        if task is None:
+            return make_response(f"Task #{task_id} not found"), 404
+        if not task.completed_at:
+            completed_at = False
+        else:
+            completed_at = task.completed_at["completed_at"]
+        return {
+            "task":{
+            "id": task.task_id,
+            "title": task.title,
+            "description": task.description,
+            "is_complete": completed_at
+            }
+        }, 200
+
+    # Update a task
+    elif request.method == "PUT":
+        if task is None:
+            return make_response(f"Task #{task_id} not found"), 404
+        form_data = request.get_json()
+
+        if "title" not in form_data or "description" not in form_data or "completed_at" not in form_data:
+            return {"details": "Invalid data"}, 400
+
+        if not "completed_at" in form_data or not form_data["completed_at"]:
+            is_complete = False
+        else:
+            is_complete = True            
+        task.title = form_data["title"]
+        task.description = form_data["description"]
+        task.completed_at = form_data["completed_at"]
+
+        db.session.commit()
+        return {
+            "task": {
+                "id": task.task_id,
+                "title": task.title,
+                "description": task.description, 
+                "is_complete": is_complete
+            }
+        }, 200
+
+    # Delete a task
+    elif request.method == "DELETE":
+        if not task:
+            return "", 404
+        db.session.delete(task)
+        db.session.commit()
+        return {
+        "details": f"Task {task_id} \"{task.title}\" successfully deleted"
+        }, 200
+            
+
+# modify complete task
+@task_bp.route("/<task_id>/mark_complete", methods=["PATCH"], strict_slashes = False)
+def mark_task_complete(task_id):
+    task = Task.query.get(task_id)
+
+    if not task: # this is the task check (wash my car)
+        return make_response(f"Task #{task_id} not found"), 404
+    task.completed_at = datetime.utcnow()
+    db.session.commit()
+
+    base_path = "https://slack.com/api/chat.postMessage"
+    slack_tocken = os.environ.get("SLACK_API_KEY")
+    query_params = {
+        "channel": "task-notifications",
+        "text": f"Someone just completed the {task.title}"
+    }
+    headers = {
+        "Authorization": f"Bearer {slack_tocken}"
+    }
+    requests.post(base_path, params=query_params, headers=headers)
```

Comment:
👍 

Nice Slack integration. We could consider moving this to a helper function so that the main purpose of this function (marking the task complete) is easier to see.

Also, since we're sending a post request, we would typically send the parameters as form data, rather than as query params.

Query params do have a maximum length (as part of the HTTP standard), so when we have potentially large data (like a text message), we often send that data in the form-encoded body of a POST request (this stems from older web standards. Now, we might use JSON in the request body).

With the `requests` library, we can set the form-encoded body by using the `data` named parameter rather than `params` (which sets the query params).

```python
    requests.post(base_path, data=query_params, headers=headers)
```